winlogbeat/eventlog/runner.go

```diff
@@ -45,65 +46,52 @@ func Run(
 	// to shut down or when returning after io.EOF
 	cancelCtx, cancelFn := ctxtool.WithFunc(ctx, func() {
 		if err := api.Close(); err != nil {
-			log.Errorw("Error while closing Windows Event Log access", "error", err)
+			log.Errorw("error while closing Windows Event Log access", "error", err)
 		}
 	})
 	defer cancelFn()
 
-	// Flag used to detect repeat "channel not found" errors, eliminating log spam.
-	channelNotFoundErrDetected := false
+	openErrHandler := newExponentialLimitedBackoff(log, 5*time.Second, time.Minute, func(err error) bool {
+		if IsRecoverable(err, api.IsFile()) {
+			log.Errorw("encountered recoverable error when opening Windows Event Log", "error", err)
+			return true
+		}
+		return false
+	})
 
-runLoop:
-	for {
-		//nolint:nilerr // only log error if we are not shutting down
-		if cancelCtx.Err() != nil {
-			return nil
+	readErrHandler := newExponentialLimitedBackoff(log, 5*time.Second, time.Minute, func(err error) bool {
+		if IsRecoverable(err, api.IsFile()) {
+			log.Errorw("encountered recoverable error when reading from Windows Event Log", "error", err)
+			if resetErr := api.Reset(); resetErr != nil {
+				log.Errorw("error resetting Windows Event Log handle", "error", resetErr)
+			}
+			return true
 		}
+		return false
+	})
 
+runLoop:
+	for cancelCtx.Err() == nil {
 		openErr := api.Open(evtCheckpoint)
-
-		switch {
-		case IsRecoverable(openErr):
-			log.Errorw("Encountered recoverable error when opening Windows Event Log", "error", openErr)
-			_ = timed.Wait(cancelCtx, 5*time.Second)
-			continue
-		case !api.IsFile() && IsChannelNotFound(openErr):
-			if !channelNotFoundErrDetected {
-				log.Errorw("Encountered channel not found error when opening Windows Event Log", "error", openErr)
-			} else {
-				log.Debugw("Encountered channel not found error when opening Windows Event Log", "error", openErr)
+		if openErr != nil {
+			if openErrHandler.backoff(cancelCtx, openErr) {
+				continue
 			}
-			channelNotFoundErrDetected = true
```

Comment:
The reason why we have this block is to suppress error logs from "channel not found" errors. I believe this arose out of a user complaint some time ago. This probably wasn't the most elegant solution, but it worked at the time. I really do like the new error handling mechanism, but we may have to reintroduce the `IsChannelNotFound` func or something similar so we can keep the existing behavior.

Reply:
as I understood previously it was spamming constantly, wouldn't this be solved now with the backoff? Other is that we do not consider "channel not found" actually recoverable in the same sense, and try at most X times if not resolved. But if not, I think it makes sense to see the logs the same as the other errors. Does that make sense?

Reply:
That's a really good point, actually. We didn't have any sort of backoff at the time, so yeah it would've just spammed until the end of time. I agree that we shouldn't hide error logs, either, and that the backoff would address the spam issue.